### ISSUE TITLE
conda has deps on nbformat, not included in miniconda

### ISSUE THREAD

**Issue by ijstokes** [conda/conda#4796]
If someone installs miniconda and then attempts to do `conda env attach` they are going to be disappointed because `nbformat` is not included in miniconda.  I don't think it should be, but then perhaps `conda env attach` should be prepared to say "sorry, you can't execute this command with just miniconda -- "conda install nbformat" if you like and try again".  The conda error page doesn't make it clear what the problem is or what the solution might be.

```
conda env attach -r ijstokes/tf_digits tensorflow.ipynb                       
Environment None will be attach into tensorflow.ipynb                                    
An unexpected error has occurred.                                                        
Please consider posting the following information to the                                 
conda GitHub issue tracker at:                                                           
                                                                                         
    https://github.com/conda/conda/issues                                                
                                                                                         
                                                                                         
                                                                                         
Current conda install:                                                                   
                                                                                         
               platform : linux-64                                                       
          conda version : 4.2.13                                                         
       conda is private : False                                                          
      conda-env version : 4.2.13                                                         
    conda-build version : not installed                                                  
         python version : 3.5.2.final.0                                                  
       requests version : 2.9.1                                                          
       root environment : /home/demo/miniconda3  (writable)                              
    default environment : /projects/demo/deeplearn/envs/dnn                              
       envs directories : /projects/demo/deeplearn/envs                                  
                          /home/demo/miniconda3/envs                                     
          package cache : /projects/demo/deeplearn/envs/.pkgs                            
                          /home/demo/miniconda3/pkgs                                     
           channel URLs : https://repo.continuum.io/pkgs/free/linux-64                   
                          https://repo.continuum.io/pkgs/free/noarch                     
                          https://repo.continuum.io/pkgs/pro/linux-64                    
                          https://repo.continuum.io/pkgs/pro/noarch                      
            config file : /projects/demo/deeplearn/.condarc                              
           offline mode : False                                                          
                                                                                         
                                                                                         
                                                                                         
`$ /home/demo/miniconda3/bin/conda-env attach -r ijstokes/tf_digits tensorflow.i         
                                                                                         
                                                                                         
                                                                                         
                                                                                         
    Traceback (most recent call last):                                                   
      File "/home/demo/miniconda3/lib/python3.5/site-packages/conda/exceptions.p         
        return_value = func(*args, **kwargs)                                             
      File "/home/demo/miniconda3/lib/python3.5/site-packages/conda_env/cli/main         
        nb = Notebook(args.notebook)                                                     
      File "/home/demo/miniconda3/lib/python3.5/site-packages/conda_env/utils/no         
        raise NBFormatNotInstalled                                                       
    conda_env.exceptions.NBFormatNotInstalled: Runtime error: nbformat is not in         
            conda install nbformat
```

**Comment by kalefranz** [comment]
We should probably deprecate `conda env attach` for the same reasons we should deprecate `conda env upload`.  `conda env upload` requires speaking the Anaconda API, which is the job of anaconda-client, not of conda.  The command should thus be `anaconda env upload` or something like that.  Just like any awareness of an Anaconda API is out of scope for conda, so is awareness of the concept of a "notebook."  The description for `conda env attach` is 

> Embeds information describing your conda environment into the notebook metadata

The dependency order is fundamentally flipped.

**Comment by ijstokes** [comment]
@kalefranz this goes back to the debate from 12-24 months ago around:

* "*conda environments*"
* "*conda env*" specifications (different from a "*conda environment*")
* "*kapsels*" (as they were then called) (a third variation), and its relationship to the emerging "*anaconda project*"

And the specific question here: how do we easily capture the conda environment required to run a Jupyter Notebook that will satisfy 80% of cases?  `conda env attach` did this but leads to this problem.  `anaconda project` provides an alternative way of doing this but I don't want to parallel every Notebook file I want to share with a project YAML file (so it fails on the *"easily"* and *"80% of cases"* front).  And consequently we're assuming users are OK with understanding the differences between:

* `conda`
* `conda env`
* `anaconda`
* `anaconda-project`

@havocp 's thoughts on this issue would be appreciated, although perhaps others from the conda leadership or even more senior need to weigh in.

I'm left feeling like a blog post trying to explain this would start off as a bad tech comedy sketch with "It's all very simple, you see ... "

**Comment by havocp** [comment]
The user workflow in Anaconda Platform as it's shaping up currently doesn't really have direct use of `conda` or `conda env` or `environment.yml` in it - nothing stops you from using them, but the beginning user docs would not need to go into them to get people started.

So we have the two commands that _are_ in the docs - at least the command line docs - `anaconda` and `anaconda-project`. The distinction between `anaconda` and `anaconda-project` is that `anaconda` affects the Anaconda platform server and `anaconda-project` affects your local project directory.

It's also possible to work on a notebook with only the GUI and no command line tools. In Jupyter there will be a "Packages" list and you can add packages to the list ... there's no manual environment manipulation. That work is currently on a branch at https://github.com/continuumio/bokeh-lab/tree/jlab19

The way we're doing this the workflow can be something like 1) click "new notebook", 2) see notebook now open, 3) click "add package" to add packages you need, 4) work on notebook, 5) click the save or share button to share with colleagues. The command line tools have equivalent operations. The "make environment", "activate environment", "install package", etc. steps are not done manually. There's also no need to manually edit `anaconda-project.yml` unless you feel like it.

This is an IDE-style experience, it won't be what everyone wants, but manual conda environment manipulation will always still be possible for those who choose to dig in.

`conda env attach` is a little bit of an orphan perhaps since it's too high-level for conda and too low-level/wrong-model for Anaconda Platform. So I don't know where it ends up. We don't have plans to do anything with it for Anaconda Platform afaik.

**Comment by kalefranz** [comment]
Should the functionality of conda env attach be part of nb_conda? And then you could attach an environment to a notebook from within the notebook, rather than run a command external to the notebook that injects information into it?

**Comment by ijstokes** [comment]
@kalefranz that is a good question and a good idea.  I like it.  It ties the "current" conda environment to the "current" notebook.  I think `nb_conda_kernels` might be a better Jupyter extension to attach it to, however, since I believe it is that extension that has knowledge of the "current" conda environment.

@havocp just to confirm: this idea of "add packages" is exactly "add conda packages in a conda manner with standard conda encoding when serialized", correct?  At some point I think there was discussion of this being "generalized" so we would have two different ideas of "packages" (and "groups of packages"), and also two different ways of serializing the definition of "a set of packages" (well, we already have two: `conda env export` and `conda list -e`, so I should probably say "a third")

**Comment by havocp** [comment]
Add packages both installs the (conda) packages and records the package name/version in the project file. environment.yml isn't used by default, nor conda env nor conda list. Those are all lower-layer "I want to do it manually" tools that don't arise when you are doing things on the project/platform layer.

I haven't heard anything about two ideas of packages though.

In the "project" layer, serializing environments isn't a user operation. Environments are always and automatically serialized. Users just add/remove packages and that affects both environment and serialized environment in one step behind the scenes.

There is an import feature to pull in an environment.yml to a project though.

**Comment by kalefranz** [comment]
Will be resolved in conda 4.6 with https://github.com/conda/conda/pull/6916
